simulator/modules/decode/decode.cpp

```diff
@@ -94,24 +98,31 @@ void Decode<FuncInstr>::clock( Cycle cycle)
         return;
     }
 
-    auto instr = read_instr( cycle);
+    bool from_stall;
+    auto instr = read_instr( cycle, from_stall);
 
     /* acquiring real information for BPU */
     wp_bp_update->write( instr.get_bp_upd(), cycle);
 
-    bool is_misprediction = instr.is_direct_jump() &&
-    ( !instr.get_bp_data().is_taken || instr.get_bp_data().target != instr.get_decoded_target());
+    if ( instr.is_jump())
+        num_jumps++;
 
     /* handle misprediction */
-    if ( is_misprediction)
+    if ( is_misprediction( instr, instr.get_bp_data()))
     {
+        num_mispredictions++;
+
         // flushing fetch stage, instr fetch will appear at decode stage next clock,
         // so we send flush signal to decode
-        wp_flush_fetch->write( true, cycle);
+        if ( !bypassing_unit->is_stall( instr))
+            wp_flush_fetch->write( true, cycle);
```

Comment:
Could you please describe the reason for that change? Time diagram would work well.

Reply:
If we detect short misprediction and simultaneously have data hazard at this stage, then decoder would flush the instruction. Not the instruction from fetch stage, but the instruction, which continues to stay on decode stage.

Reply:
I suppose we can just swap the order of flush and stall checks, so that decode stage wouldn't reach 'flush' step if it has 'stall'.

Reply:
Let's do that in the next pull request.